data cleaning 

1.import necessary packages 

In [25]:
#imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('ggplot')
import warnings
warnings.filterwarnings('ignore')
# from pandas_profiling import ProfileReport
# pd.set_option('display.max_rows', None)
# pd.set_option('display.max_columns', None)

%matplotlib inline

In [29]:
#read csv
solar_ctry = pd.read_csv('../data/EMHIRESPV_TSh_CF_Country_19862015.csv')
solar_nuts = pd.read_csv('../data/EMHIRES_PVGIS_TSh_CF_n2_19862015.csv')

#preview data
print("---------------------------- Top 5 rows of solar_ctry -------------------------")
display(solar_ctry.head())
print(f'Shape of solar_ctry : {solar_ctry.shape}')
print("----------------------------- Data types in solar_ctry ------------------------")
print(solar_ctry.dtypes)
print("---------------------------- Top 5 rows of solar_nuts -------------------------")
display(solar_nuts.head())
print(f'Shape of solar_nuts : {solar_nuts.shape}')
print("----------------------------- Data types in solar_nuts ------------------------")
print(solar_nuts.dtypes)

---------------------------- Top 5 rows of solar_ctry -------------------------


AT   BE   BG   CH  CY   CZ   DE   DK   EE   ES  ...   LV   NL   NO   PL  \
0  0.0  0.0  0.0  0.0   0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0   0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0   0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0   0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
4  0.0  0.0  0.0  0.0   0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

    PT   RO   SI   SK   SE   UK  
0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 29 columns]

Shape of solar_ctry : (262968, 29)
----------------------------- Data types in solar_ctry ------------------------
AT    float64
BE    float64
BG    float64
CH    float64
CY      int64
CZ    float64
DE    float64
DK    float64
EE    float64
ES    float64
FI    float64
FR    float64
EL    float64
HR    float64
HU    float64
IE    float64
IT    float64
LT    float64
LU    float64
LV    float64
NL    float64
NO    float64
PL    float64
PT    float64
RO    float64
SI    float64
SK    float64
SE    float64
UK    float64
dtype: object
---------------------------- Top 5 rows of solar_nuts -------------------------


time_step  AT11  AT21  AT12  AT31  AT32  AT22  AT33  AT34  AT13  ...  UKC2  \
0          1   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1          2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2          3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
3          4   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
4          5   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

   UKI5UKI6  UKG2  UKM3  UKE3  UKJ2  UKC1  UKG3  UKL1  UKE4  
0       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 261 columns]

Shape of solar_nuts : (262968, 261)
----------------------------- Data types in solar_nuts ------------------------
time_step      int64
AT11         float64
AT21         float64
AT12         float64
AT31         float64
              ...   
UKJ2         float64
UKC1         float64
UKG3         float64
UKL1         float64
UKE4         float64
Length: 261, dtype: object


## Overview of Datasets

### `solar_ctry`
- **Columns**: Represent various European countries, such as:
  - `AT` for Austria, `BE` for Belgium, `BG` for Bulgaria, etc.
- **Rows**: Each row represents data for a specific time step or observation period.
- **Values**: The values in the columns are of type `float64` and likely represent solar-related metrics, such as:
  - Solar power generation, consumption, or other solar-related statistics for each country over time.
  - A value of `0.0` in most rows indicates no data or zero solar output for that country at that time.

#### Example Interpretation:
- `AT` (Austria) column shows solar-related data for Austria at various time steps.
- A `0.0` value indicates no solar output during that specific time.

#### Shape: 
- Rows: 262,968
- Columns: 29

---

### `solar_nuts`
- **Columns**:
  - `time_step`: Indicates the time step or observation period for each row (likely sequential or chronological).
  - Region-specific columns, such as `AT11`, `AT21`, `AT12`, `AT31`, etc., represent specific regions or datasets within a country.
  - For example:
    - `AT11` could represent a specific region in Austria.
    - `UKC1` could represent a region or dataset for the UK.
- **Values**: The columns contain `float64` values, likely representing solar energy metrics (e.g., energy generation or consumption) for specific regions at specific time steps.

#### Example Interpretation:
- The data for each country is divided by region (e.g., `AT11`, `AT21` for Austria) to capture regional solar energy statistics.
- A value of `0.0` typically indicates no data or zero solar output for that region at a given time.

#### Shape:
- Rows: 262,968
- Columns: 261


In [47]:
def add_time(df):
    '''adds time column based on start and end year'''
    df['time'] = pd.date_range(start='1/1/1986', periods=df.shape[0],freq='H')
    df['hour'] = df['time'].dt.hour
    df['week'] = df['time'].dt.isocalendar().week  # Updated to use isocalendar
    df['month'] = df['time'].dt.month
    #check
    print("-------------- Top 5 rows ----------------")
    display(df.head())
    print("------------------ Last 5 rows -----------")
    display(df.tail())

In [35]:
country_name = {
    'AT': 'Austria',
    'BE': 'Belgium',
    'BG': 'Bulgaria',
    'CH': 'Switzerland',
    'CY': 'Cyprus',
    'CZ': 'Czech Republic',
    'DE': 'Germany',
    'DK': 'Denmark',
    'EE': 'Estonia',
    'ES': 'Spain',
    'FI': 'Finland',
    'FR': 'France',
    'EL': 'Greece',
    'HR': 'Croatia',
    'HU': 'Hungary',
    'IE': 'Ireland',
    'IT': 'Italy',
    'LT': 'Lithuania',
    'LU': 'Luxembourg',
    'LV': 'Latvia',
    'NL': 'Netherlands',
    'NO': 'Norway',
    'PL': 'Poland',
    'PT': 'Portugal',
    'RO': 'Romania',
    'SI': 'Slovenia',
    'SK': 'Slovakia',
    'SE': 'Sweden',
    'UK': 'United Kingdom'
}

solar_ctry.columns = [country_name.get(col, col) for col in solar_ctry .columns]


In [37]:
solar_ctry 

Austria  Belgium  Bulgaria  Switzerland  Cyprus  Czech Republic  \
0           0.0      0.0       0.0          0.0       0             0.0   
1           0.0      0.0       0.0          0.0       0             0.0   
2           0.0      0.0       0.0          0.0       0             0.0   
3           0.0      0.0       0.0          0.0       0             0.0   
4           0.0      0.0       0.0          0.0       0             0.0   
...         ...      ...       ...          ...     ...             ...   
262963      0.0      0.0       0.0          0.0       0             0.0   
262964      0.0      0.0       0.0          0.0       0             0.0   
262965      0.0      0.0       0.0          0.0       0             0.0   
262966      0.0      0.0       0.0          0.0       0             0.0   
262967      0.0      0.0       0.0          0.0       0             0.0   

        Germany  Denmark  Estonia  Spain  ...  Latvia  Netherlands  Norway  \
0           0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   
1           0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   
2           0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   
3           0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   
4           0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   
...         ...      ...      ...    ...  ...     ...          ...     ...   
262963      0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   
262964      0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   
262965      0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   
262966      0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   
262967      0.0      0.0      0.0    0.0  ...     0.0          0.0     0.0   

        Poland  Portugal  Romania  Slovenia  Slovakia  Sweden  United Kingdom  
0          0.0       0.0      0.0       0.0       0.0     0.0             0.0  
1          0.0       0.0      0.0       0.0       0.0     0.0             0.0  
2          0.0       0.0      0.0       0.0       0.0     0.0             0.0  
3          0.0       0.0      0.0       0.0       0.0     0.0             0.0  
4          0.0       0.0      0.0       0.0       0.0     0.0             0.0  
...        ...       ...      ...       ...       ...     ...             ...  
262963     0.0       0.0      0.0       0.0       0.0     0.0             0.0  
262964     0.0       0.0      0.0       0.0       0.0     0.0             0.0  
262965     0.0       0.0      0.0       0.0       0.0     0.0             0.0  
262966     0.0       0.0      0.0       0.0       0.0     0.0             0.0  
262967     0.0       0.0      0.0       0.0       0.0     0.0             0.0  

[262968 rows x 29 columns]

In [49]:
#apply function that adds time columns and its variants
add_time(solar_ctry)
add_time(solar_nuts)

-------------- Top 5 rows ----------------


Austria  Belgium  Bulgaria  Switzerland  Cyprus  Czech Republic  Germany  \
0      0.0      0.0       0.0          0.0       0             0.0      0.0   
1      0.0      0.0       0.0          0.0       0             0.0      0.0   
2      0.0      0.0       0.0          0.0       0             0.0      0.0   
3      0.0      0.0       0.0          0.0       0             0.0      0.0   
4      0.0      0.0       0.0          0.0       0             0.0      0.0   

   Denmark  Estonia  Spain  ...  Portugal  Romania  Slovenia  Slovakia  \
0      0.0      0.0    0.0  ...       0.0      0.0       0.0       0.0   
1      0.0      0.0    0.0  ...       0.0      0.0       0.0       0.0   
2      0.0      0.0    0.0  ...       0.0      0.0       0.0       0.0   
3      0.0      0.0    0.0  ...       0.0      0.0       0.0       0.0   
4      0.0      0.0    0.0  ...       0.0      0.0       0.0       0.0   

   Sweden  United Kingdom                time  hour  week  month  
0     0.0             0.0 1986-01-01 00:00:00     0     1      1  
1     0.0             0.0 1986-01-01 01:00:00     1     1      1  
2     0.0             0.0 1986-01-01 02:00:00     2     1      1  
3     0.0             0.0 1986-01-01 03:00:00     3     1      1  
4     0.0             0.0 1986-01-01 04:00:00     4     1      1  

[5 rows x 33 columns]

------------------ Last 5 rows -----------


Austria  Belgium  Bulgaria  Switzerland  Cyprus  Czech Republic  \
262963      0.0      0.0       0.0          0.0       0             0.0   
262964      0.0      0.0       0.0          0.0       0             0.0   
262965      0.0      0.0       0.0          0.0       0             0.0   
262966      0.0      0.0       0.0          0.0       0             0.0   
262967      0.0      0.0       0.0          0.0       0             0.0   

        Germany  Denmark  Estonia  Spain  ...  Portugal  Romania  Slovenia  \
262963      0.0      0.0      0.0    0.0  ...       0.0      0.0       0.0   
262964      0.0      0.0      0.0    0.0  ...       0.0      0.0       0.0   
262965      0.0      0.0      0.0    0.0  ...       0.0      0.0       0.0   
262966      0.0      0.0      0.0    0.0  ...       0.0      0.0       0.0   
262967      0.0      0.0      0.0    0.0  ...       0.0      0.0       0.0   

        Slovakia  Sweden  United Kingdom                time  hour  week  \
262963       0.0     0.0             0.0 2015-12-31 19:00:00    19    53   
262964       0.0     0.0             0.0 2015-12-31 20:00:00    20    53   
262965       0.0     0.0             0.0 2015-12-31 21:00:00    21    53   
262966       0.0     0.0             0.0 2015-12-31 22:00:00    22    53   
262967       0.0     0.0             0.0 2015-12-31 23:00:00    23    53   

        month  
262963     12  
262964     12  
262965     12  
262966     12  
262967     12  

[5 rows x 33 columns]

-------------- Top 5 rows ----------------


time_step  AT11  AT21  AT12  AT31  AT32  AT22  AT33  AT34  AT13  ...  UKE3  \
0          1   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1          2   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2          3   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
3          4   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
4          5   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

   UKJ2  UKC1  UKG3  UKL1  UKE4                time  hour  week  month  
0   0.0   0.0   0.0   0.0   0.0 1986-01-01 00:00:00     0     1      1  
1   0.0   0.0   0.0   0.0   0.0 1986-01-01 01:00:00     1     1      1  
2   0.0   0.0   0.0   0.0   0.0 1986-01-01 02:00:00     2     1      1  
3   0.0   0.0   0.0   0.0   0.0 1986-01-01 03:00:00     3     1      1  
4   0.0   0.0   0.0   0.0   0.0 1986-01-01 04:00:00     4     1      1  

[5 rows x 265 columns]

------------------ Last 5 rows -----------


time_step  AT11  AT21  AT12  AT31  AT32  AT22  AT33  AT34  AT13  ...  \
262963     262964   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
262964     262965   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
262965     262966   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
262966     262967   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
262967     262968   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   

        UKE3  UKJ2  UKC1  UKG3  UKL1  UKE4                time  hour  week  \
262963   0.0   0.0   0.0   0.0   0.0   0.0 2015-12-31 19:00:00    19    53   
262964   0.0   0.0   0.0   0.0   0.0   0.0 2015-12-31 20:00:00    20    53   
262965   0.0   0.0   0.0   0.0   0.0   0.0 2015-12-31 21:00:00    21    53   
262966   0.0   0.0   0.0   0.0   0.0   0.0 2015-12-31 22:00:00    22    53   
262967   0.0   0.0   0.0   0.0   0.0   0.0 2015-12-31 23:00:00    23    53   

        month  
262963     12  
262964     12  
262965     12  
262966     12  
262967     12  

[5 rows x 265 columns]

In [77]:

# Check for missing values
print("-------------- Missing Values in Solar_ctry ----------------")
print(solar_ctry.isnull().sum())

print("-------------- Missing Values in Solar_nuts ----------------")
print(solar_nuts.isnull().sum())



-------------- Missing Values in Solar_ctry ----------------
Austria           0
Belgium           0
Bulgaria          0
Switzerland       0
Cyprus            0
Czech Republic    0
Germany           0
Denmark           0
Estonia           0
Spain             0
Finland           0
France            0
Greece            0
Croatia           0
Hungary           0
Ireland           0
Italy             0
Lithuania         0
Luxembourg        0
Latvia            0
Netherlands       0
Norway            0
Poland            0
Portugal          0
Romania           0
Slovenia          0
Slovakia          0
Sweden            0
United Kingdom    0
time              0
hour              0
week              0
month             0
dtype: int64
-------------- Missing Values in Solar_nuts ----------------
time_step    0
AT11         0
AT21         0
AT12         0
AT31         0
            ..
UKE4         0
time         0
hour         0
week         0
month        0
Length: 265, dtype: int64


Both datasets (`solar_ctry` and `solar_nuts`) do not contain any missing values, making them ready for further analysis or processing.

In [80]:
solar_ctry.to_csv('../data/solar-ctry-clean.csv', index = False)
solar_nuts.to_csv('../data/solar-nuts-clean.csv', index = False)